File: HMCLCore/src/main/java/org/jackhuang/hmcl/task/FileDownloadTask.java
/*
 * Hello Minecraft! Launcher
 * Copyright (C) 2020  huangyuhui <huanghongxun2008@126.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.jackhuang.hmcl.task;

import org.jackhuang.hmcl.util.DigestUtils;
import org.jackhuang.hmcl.util.io.ChecksumMismatchException;
import org.jackhuang.hmcl.util.io.CompressingUtils;
import org.jackhuang.hmcl.util.io.FileUtils;
import org.jackhuang.hmcl.util.io.NetworkUtils;

import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.net.http.HttpResponse;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.util.*;

import static java.util.Objects.requireNonNull;
import static org.jackhuang.hmcl.util.logging.Logger.LOG;

/**
 * A task that can download a file online.
 *
 * @author huangyuhui
 */
public class FileDownloadTask extends FetchTask<Void> {

    public static class IntegrityCheck {
        private final String algorithm;
        private final String checksum;

        public IntegrityCheck(String algorithm, String checksum) {
            this.algorithm = requireNonNull(algorithm);
            this.checksum = requireNonNull(checksum);
        }

        public static IntegrityCheck of(String algorithm, String checksum) {
            if (checksum == null) return null;
            else return new IntegrityCheck(algorithm, checksum);
        }

        public String getAlgorithm() {
            return algorithm;
        }

        public String getChecksum() {
            return checksum;
        }

        @Override
        public String toString() {
            return String.format("IntegrityCheck[algorithm='%s', checksum='%s']", algorithm, checksum);
        }
    }

    private final Path file;
    private final IntegrityCheck integrityCheck;
    private boolean caching;
    private Path candidate;
    private final ArrayList<IntegrityCheckHandler> integrityCheckHandlers = new ArrayList<>();

    /**
     * @param uri  the URI of remote file.
     * @param path the location that download to.
     */
    public FileDownloadTask(String uri, Path path) {
        this(List.of(NetworkUtils.toURI(uri)), path, null);
    }

    /**
     * @param uri            the URI of remote file.
     * @param path           the location that download to.
     * @param integrityCheck the integrity check to perform, null if no integrity check is to be performed
     */
    public FileDownloadTask(String uri, Path path, IntegrityCheck integrityCheck) {
        this(List.of(NetworkUtils.toURI(uri)), path, integrityCheck);
    }

    /**
     * @param uri  the URI of remote file.
     * @param path the location that download to.
     */
    public FileDownloadTask(URI uri, Path path) {
        this(uri, path, null);
    }

    /**
     * @param uri            the URI of remote file.
     * @param path           the location that download to.
     * @param integrityCheck the integrity check to perform, null if no integrity check is to be performed
     */
    public FileDownloadTask(URI uri, Path path, IntegrityCheck integrityCheck) {
        this(List.of(uri), path, integrityCheck);
    }

    /**
     * Constructor.
     *
     * @param uris uris of remote file, will be attempted in order.
     * @param file the location that download to.
     */
    public FileDownloadTask(List<URI> uris, Path file) {
        this(uris, file, null);
    }

    /**
     * Constructor.
     *
     * @param uris           uris of remote file, will be attempted in order.
     * @param path           the location that download to.
     * @param integrityCheck the integrity check to perform, null if no integrity check is to be performed
     */
    public FileDownloadTask(List<URI> uris, Path path, IntegrityCheck integrityCheck) {
        super(uris);
        this.file = path;
        this.integrityCheck = integrityCheck;

        setName(path.getFileName().toString());
    }

    public Path getPath() {
        return file;
    }

    public void setCaching(boolean caching) {
        this.caching = caching;
    }

    public FileDownloadTask setCandidate(Path candidate) {
        this.candidate = candidate;
        return this;
    }

    public void addIntegrityCheckHandler(IntegrityCheckHandler handler) {
        integrityCheckHandlers.add(Objects.requireNonNull(handler));
    }

    @Override
    protected EnumCheckETag shouldCheckETag() {
        // Check cache
        if (integrityCheck != null && caching) {
            Optional<Path> cache = repository.checkExistentFile(candidate, integrityCheck.getAlgorithm(), integrityCheck.getChecksum());
            if (cache.isPresent()) {
                try {
                    FileUtils.copyFile(cache.get(), file);
                    LOG.trace("Successfully verified file " + file + " from " + uris.get(0));
                    return EnumCheckETag.CACHED;
                } catch (IOException e) {
                    LOG.warning("Failed to copy cache files", e);
                }
            }
            return EnumCheckETag.NOT_CHECK_E_TAG;
        } else {
            return EnumCheckETag.CHECK_E_TAG;
        }
    }

    @Override
    protected void beforeDownload(URI uri) {
        LOG.trace("Downloading " + uri + " to " + file);
    }

    @Override
    protected void useCachedResult(Path cache) throws IOException {
        FileUtils.copyFile(cache, file);
    }

    @Override
    protected Context getContext(HttpResponse<?> response, boolean checkETag, String bmclapiHash) throws IOException {
        Path temp = Files.createTempFile(null, null);

        String algorithm;
        String checksum;
        if (integrityCheck != null) {
            algorithm = integrityCheck.getAlgorithm();
            checksum = integrityCheck.getChecksum();
        } else if (bmclapiHash != null && DigestUtils.isSha1Digest(bmclapiHash)) {
            algorithm = "SHA-1";
            checksum = bmclapiHash;
        } else {
            algorithm = null;
            checksum = null;
        }

        MessageDigest digest = algorithm != null ? DigestUtils.getDigest(algorithm) : null;

        OutputStream fileOutput = Files.newOutputStream(temp);
        return new Context() {
            @Override
            public void write(byte[] buffer, int offset, int len) throws IOException {
                if (digest != null) {
                    digest.update(buffer, offset, len);
                }

                fileOutput.write(buffer, offset, len);
            }

            @Override
            public void close() throws IOException {
                try {
                    fileOutput.close();
                } catch (IOException e) {
                    LOG.warning("Failed to close file: " + temp, e);
                }

                if (!isSuccess()) {
                    try {
                        Files.deleteIfExists(temp);
                    } catch (IOException e) {
                        LOG.warning("Failed to delete file: " + temp, e);
                    }
                    return;
                }

                for (IntegrityCheckHandler handler : integrityCheckHandlers) {
                    handler.checkIntegrity(temp, file);
                }

                Files.createDirectories(file.toAbsolutePath().getParent());

                try {
                    Files.move(temp, file, StandardCopyOption.REPLACE_EXISTING);
                } catch (Exception e) {
                    throw new IOException("Unable to move temp file from " + temp + " to " + file, e);
                }

                // Integrity check
                if (checksum != null) {
                    String actualChecksum = HexFormat.of().formatHex(digest.digest());
                    if (!checksum.equalsIgnoreCase(actualChecksum)) {
                        throw new ChecksumMismatchException(algorithm, checksum, actualChecksum);
                    }
                }

                if (caching && algorithm != null) {
                    try {
                        repository.cacheFile(file, algorithm, checksum);
                    } catch (IOException e) {
                        LOG.warning("Failed to cache file", e);
                    }
                }

                if (checkETag) {
                    repository.cacheRemoteFile(response, file);
                }
            }
        };
    }

    public interface IntegrityCheckHandler {
        /**
         * Check whether the file is corrupted or not.
         *
         * @param filePath        the file locates in (maybe in temp directory)
         * @param destinationPath for real file name
         * @throws IOException if the file is corrupted
         */
        void checkIntegrity(Path filePath, Path destinationPath) throws IOException;
    }

    public static final IntegrityCheckHandler ZIP_INTEGRITY_CHECK_HANDLER = (filePath, destinationPath) -> {
        String ext = FileUtils.getExtension(destinationPath).toLowerCase(Locale.ROOT);
        if (ext.equals("zip") || ext.equals("jar")) {
            try (FileSystem ignored = CompressingUtils.createReadOnlyZipFileSystem(filePath)) {
                // test for zip format
            }
        }
    };
}

File: HMCLCore/src/main/kotlin/org/jackhuang/hmcl/task/FileDownloadTask.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.task

import org.jackhuang.hmcl.event.EventManager
import org.jackhuang.hmcl.event.FailedEvent
import org.jackhuang.hmcl.util.*
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.RandomAccessFile
import java.math.BigInteger
import java.net.Proxy
import java.net.URL
import java.util.logging.Level

/**
 * A task that can download a file online.
 *
 * @param url the URL of remote file.
 * @param file the location that download to.
 * @param hash the SHA-1 hash code of remote file, null if the hash is unknown or it is no need to check the hash code.
 * @param retry the times for retrying if downloading fails.
 * @param proxy the proxy.
 *
 * @author huangyuhui
 */
class FileDownloadTask @JvmOverloads constructor(val url: URL, val file: File, val hash: String? = null, val retry: Int = 5, val proxy: Proxy = Proxy.NO_PROXY): Task() {
    override val scheduler: Scheduler = Scheduler.IO

    /**
     * Once downloading fails, this event will be fired to gain the substitute URL.
     */
    var onFailed = EventManager<FailedEvent<URL>>()

    private var rFile: RandomAccessFile? = null
    private var stream: InputStream? = null

    private fun closeFiles() {
        rFile?.closeQuietly()
        rFile = null
        stream?.closeQuietly()
        stream = null
    }

    override fun execute() {
        var currentURL = url
        LOG.finer("Downloading: $currentURL, to: $file")
        var exception: Exception? = null
        for (repeat in 0 until retry) {
            if (repeat > 0) {
                val event = FailedEvent(this, repeat, currentURL)
                onFailed(event)
                if (currentURL != event.newResult) {
                    LOG.fine("Switch from: $currentURL to: ${event.newResult}")
                    currentURL = event.newResult
                }
            }
            if (Thread.interrupted()) {
                Thread.currentThread().interrupt()
                break
            }

            var temp: File? = null

            try {
                updateProgress(0.0)

                val conn = url.createConnection(proxy)
                conn.connect()

                if (conn.responseCode / 100 != 2)
                    throw IOException("Server error, response code: ${conn.responseCode}")

                val contentLength = conn.contentLength
                if (contentLength < 1)
                    throw IOException("The content length is invalid")

                if (!file.absoluteFile.parentFile.makeDirectory())
                    throw IOException("Could not make directory: ${file.absoluteFile.parent}")

                temp = createTempFile("HMCLCore")
                rFile = RandomAccessFile(temp, "rw")

                val digest = DigestUtils.sha1Digest

                stream = conn.inputStream
                var lastDownloaded = 0
                var downloaded = 0
                var lastTime = System.currentTimeMillis()
                val buf = ByteArray(4096)
                while (true) {
                    if (Thread.interrupted()) {
                        Thread.currentThread().interrupt()
                        break
                    }

                    val read = stream!!.read(buf)
                    if (read == -1)
                        break

                    if (hash != null)
                        digest.update(buf, 0, read)

                    rFile!!.write(buf, 0, read)
                    downloaded += read

                    updateProgress(downloaded, contentLength)
                    val now = System.currentTimeMillis()
                    if (now - lastTime >= 1000L) {
                        updateMessage(((downloaded - lastDownloaded) / 1024).toString() + "KB/s")
                        lastDownloaded = downloaded
                        lastTime = now
                    }
                }

                closeFiles()

                if (Thread.interrupted()) {
                    temp.delete()
                    Thread.currentThread().interrupt()
                    break
                } else {
                    if (file.exists())
                        file.delete()
                    if (!file.absoluteFile.parentFile.makeDirectory())
                        throw IOException("Cannot make parent directory $file")
                    if (!temp.renameTo(file))
                        throw IOException("Cannot move temp file to $file")
                }

                check(downloaded == contentLength, { "Unexpected file size: $downloaded, expected: $contentLength" })

                if (hash != null) {
                    val hashCode = String.format("%1$040x", BigInteger(1, digest.digest()))
                    check(hash.equals(hashCode, ignoreCase = true), { "Unexpected hash code: $hashCode, expected: $hash" })
                }

                return
            } catch(e: Exception) {
                temp?.delete()
                exception = e
                LOG.log(Level.WARNING, "Unable to download file $currentURL", e)
            } finally {
                closeFiles()
            }
        }

        if (exception != null)
            throw exception
    }
}